psd/imxrt/psd.c

```diff
@@ -0,0 +1,214 @@
+/*
+ * Phoenix-RTOS
+ *
+ * psd - Serial Download Protocol client
+ *
+ * Copyright 2019 Phoenix Systems
+ * Author: Bartosz Ciesla, Pawel Pisarczyk, Hubert Buczynski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+//#include <usbclient.h>
+
+#include <stdio.h>
+#include <errno.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <sys/msg.h>
+#include <sys/stat.h>
+#include <sys/file.h>
+#include <sys/mman.h>
+#include <sys/types.h>
+#include <sys/platform.h>
+#include <sys/reboot.h>
+
+#include <phoenix/arch/imxrt.h>
+
+#include "../common/hid.h"
+#include "../common/sdp.h"
+
+
+#define SET_OPEN_HAB(b) (b)[0]=3;(b)[1]=0x56;(b)[2]=0x78;(b)[3]=0x78;(b)[4]=0x56;
+#define SET_CLOSED_HAB(b) (b)[0]=3;(b)[1]=0x12;(b)[2]=0x34;(b)[3]=0x34;(b)[4]=0x12;
+#define SET_COMPLETE(b) (b)[0]=4;(b)[1]=0x12;(b)[2]=0x8a;(b)[3]=0x8a;(b)[4]=0x12;
+#define SET_FILE_COMPLETE(b) (b)[0]=4;(b)[1]=0x88;(b)[2]=0x88;(b)[3]=0x88;(b)[4]=0x88;
+#define SET_HAB_ERROR(b, err) (b)[0]=4;(b)[1]=err;(b)[2]=0xaa;b[3]=0xaa;(b)[4]=0xaa;
+
+
+#define HID_REPORT_1_SIZE sizeof(sdp_cmd_t) + 1
+#define HID_REPORT_2_SIZE 1025
+#define HID_REPORT_3_SIZE 5
+#define HID_REPORT_4_SIZE 65
+
+
+#define LOG(str, ...) do { if (1) fprintf(stderr, "psd: " str "\n", ##__VA_ARGS__); } while (0)
+#define LOG_ERROR(str, ...) do { fprintf(stderr, __FILE__  ":%d error: " str "\n", __LINE__, ##__VA_ARGS__); } while (0)
+
+
+struct {
+	int run;
+	char buff[SIZE_PAGE];
+
+	int (*rf)(int, char *, unsigned int, char **);
+	int (*sf)(int, const char *, unsigned int);
+
+} psd;
+
+
+const hid_dev_setup_t hid_setup = {
+	 .ddev = {
+		.len = sizeof(usbclient_desc_dev_t), .desc_type = USBCLIENT_DESC_TYPE_DEV, .bcd_usb = 0x200,
+		.dev_class = 0, .dev_subclass = 0, .dev_prot = 0, .max_pkt_sz0 = 64,
+		.vend_id = 0x1FC9, .prod_id = 0x0135, .bcd_dev = 0x0001,
+		.man_str = 0, .prod_str = 0, .sn_str = 0,
+		.num_conf = 1
+	},
+	.dstrman = {
+		.len = 27 * 2 + 2,
+		.desc_type = USBCLIENT_DESC_TYPE_STR,
+		.string = { 'F', 0, 'r', 0, 'e', 0, 'e', 0, 's', 0, 'c', 0, 'a', 0, 'l', 0, 'e', 0, ' ', 0, 'S', 0, 'e', 0, 'm', 0, 'i', 0, 'C', 0, 'o', 0, 'n', 0, 'd', 0, 'u', 0, 'c', 0, 't', 0, 'o', 0, 'r', 0, ' ', 0, 'I', 0, 'n', 0, 'c', 0 }
+	},
+	.dstrprod = {
+		.len = 11 * 2 + 2,
+		.desc_type = USBCLIENT_DESC_TYPE_STR,
+		.string = { 'S', 0, 'E', 0, ' ', 0, 'B', 0, 'l', 0, 'a', 0, 'n', 0, 'k', 0, ' ', 0, 'R', 0, 'T', 0 }
+	}
+};
+
+
+int psd_hidResponse(int err, int type)
```

Comment:
Is err flag used anywhere at all? err and !err cases have no common code, consider splitting this function in to two functions.